### PR TITLE
ci: Use containerized services in Travis

### PR DESCRIPTION
**Pull request by adarnimrod**
Try to improve the time it takes for a ci run by:
- Caching Docker images.
- Doing less for the lint job.

Don't merge this yet, the PR is for running the CI.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,46 +1,43 @@
...
-- sudo -u postgres createdb --owner=azafea azafea-tests
+- docker build --build-arg build_type=dev --tag azafea-tests --pull .
 
-# Set the default passwords for PostgreSQL and Redis, as that's what the tests expect
-- sudo -u postgres psql --command="ALTER USER azafea WITH PASSWORD 'CHANGE ME!!';"
-- redis-cli config set requirepass 'CHANGE ME!!'
+before_cache:
+# Save Docker images to Travis' cache, but prune old ones first
+- mkdir -p docker_images
+- docker images --quiet --filter=dangling=true | xargs -r docker rmi
```
Comment (bochecha) [inline .travis.yml]: Is this the same as `docker system prune`?

File: .travis.yml [hunk 2]
```diff
@@ -1,46 +1,43 @@
...
+- docker save $(docker images -a -q) | gzip > docker_images/images.tar.gz
 
 matrix:
   include:
   - name: "Integration Tests"
+    before_script:
```
Comment (bochecha) [inline .travis.yml]: I like that you're moving these here, because they are not necessary for the "lint" job and so that one can be faster. :+1:

Can I ask you to do that in a separate commit though, so it's easier to review and understand?

(i.e first move things here then do the caching stuff, or the contrary as you prefer)

File: .travis.yml [hunk 3]
```diff
@@ -1,46 +1,43 @@
...
+    - docker image pull postgres:11-alpine
+    - docker image pull redis:5-alpine
```
Comment (bochecha) [inline .travis.yml]: You use these ones because they are smaller? (that's fine, I just want to make sure I understand the "why")

File: .travis.yml [hunk 4]
```diff
@@ -1,46 +1,43 @@
...
+- docker build --build-arg build_type=dev --tag azafea-tests --pull .
 
-# Set the default passwords for PostgreSQL and Redis, as that's what the tests expect
-- sudo -u postgres psql --command="ALTER USER azafea WITH PASSWORD 'CHANGE ME!!';"
-- redis-cli config set requirepass 'CHANGE ME!!'
+before_cache:
+# Save Docker images to Travis' cache, but prune old ones first
+- mkdir -p docker_images
+- docker images --quiet --filter=dangling=true | xargs -r docker rmi
+- docker save $(docker images -a -q) | gzip > docker_images/images.tar.gz
```
Comment (bochecha) [inline .travis.yml]: I kind of wish you'd separate the speed up through caching from the move to Docker images.

(if you're not comfortable enough with Git to split and reorder commits, I'm happy to do it for you :slightly_smiling_face: )

File: .travis.yml [hunk 5]
```diff
@@ -1,46 +1,43 @@
...
   - name: "Integration Tests"
+    before_script:
+    - docker image pull postgres:11-alpine
+    - docker image pull redis:5-alpine
+    - docker run -d --name=redis --network=host redis:5-alpine --requirepass 'CHANGE ME!!'
+    - docker run -d --name=postgres --network=host -e 'POSTGRES_PASSWORD=CHANGE ME!!' -e 'POSTGRES_USER=azafea' -e 'POSTGRES_DB=azafea-tests' postgres:11-alpine
     script:
     - docker run --rm --network=host --entrypoint="" azafea-tests pipenv run test-all
+    after_script:
+    - docker rm -f redis postgres
```
Comment (bochecha) [inline .travis.yml]: Why remove the containers? Won't that make the images dangling, and as such they'll get removed in `before_cache` above?

File: .travis.yml (hunk 1 above)
Comment (adarnimrod) [reply .travis.yml]: Yeah, but this also works on older version of Docker. Doesn't matter anymore, I removed the whole cache thing (it made the build even slower).

File: .travis.yml (hunk 3 above)
Comment (adarnimrod) [reply .travis.yml]: Just for the smaller images (faster to download). There are actually tags for the latest Alpine variants. But this could be surprising (builds could suddenly fail if the image in Docker Hub is updated to a new major version).

File: .travis.yml (hunk 5 above)
Comment (adarnimrod) [reply .travis.yml]: You don't have to remove the containers (the VM is destroyed anyway). But the images from which the containers are made aren't touched.

**Comment by adarnimrod** [comment]
I changed the PR from containers and speed up to just containers (and a little bit of a speed up). Only question I have is do you wish to the latest available version of Redis and Postgres or set versions?

File: .travis.yml (hunk 1 above)
Comment (bochecha) [reply .travis.yml]: Ah ok, thanks for the tip about older Docker then. :slightly_smiling_face:

File: .travis.yml (hunk 3 above)
Comment (bochecha) [reply .travis.yml]: > There are actually tags for the latest Alpine variants

Ah, they are caled "alpine"! I missed them because I was searching for something with "latest" in the name.

> But this could be surprising (builds could suddenly fail if the image in Docker Hub is updated to a new major version).

In general I agree with that.

For things like Redis and PostgreSQL though, I'm pretty confident about their consistent stability and compatibility that I'd be willing to risk it.

Although I guess it's nice to test with what is in production right now as well… :thinking: 

I know we use PostgreSQL 11 in production, are we also using Redis 5?

Comment (adarnimrod) [reply .travis.yml]: We're on Redis 5.0. Since for now we're the only users of this we can set it to what Endless has in production. But for a more broad appeal we can actually run against multiple versions of PG and/or Redis (whatever we decide is supported). Personally I prefer to be explicit in this regard. I imagine that when we decide to update PG or Redis in production we'll start by adding a job to test on both versions (current and future) and keep that until all environments are updated to the new version and then we'll remove the old version from the CI.

File: .travis.yml (hunk 2 above)
Comment (adarnimrod) [reply .travis.yml]: Done.

File: .travis.yml (hunk 4 above)
Comment (adarnimrod) [reply .travis.yml]: Done.

File: .travis.yml (hunk 3 above)
Comment (bochecha) [reply .travis.yml]: > Since for now we're the only users of this we can set it to what Endless has in production.

Yes, let's.
